### ISSUE TITLE
C++17 is now required in order to build the tools with up-to-date dependencies

### ISSUE THREAD

**Issue by davidebeatrici** [Tencent/ncnn#4567]
<details>
<summary>Build log</summary>

```c++
In file included from ncnn/tools/caffe/caffe2ncnn.cpp:19:
In file included from ncnn/build/tools/caffe/caffe.pb.h:11:
/usr/include/google/protobuf/port_def.inc:200:1: error: static assertion failed due to requirement '201103L >= 201402L': Protobuf only supports C++14 and newer.
static_assert(PROTOBUF_CPLUSPLUS_MIN(201402L), "Protobuf only supports C++14 and newer.");
^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ncnn/tools/caffe/caffe2ncnn.cpp:19:
In file included from ncnn/build/tools/caffe/caffe.pb.h:24:
In file included from /usr/include/google/protobuf/io/coded_stream.h:130:
In file included from /usr/include/google/protobuf/stubs/common.h:44:
In file included from /usr/include/absl/strings/string_view.h:39:
In file included from /usr/include/absl/base/attributes.h:37:
In file included from /usr/include/absl/base/config.h:86:
/usr/include/absl/base/policy_checks.h:79:2: error: "C++ versions less than C++14 are not supported."
#error "C++ versions less than C++14 are not supported."
 ^
In file included from ncnn/tools/caffe/caffe2ncnn.cpp:19:
In file included from ncnn/build/tools/caffe/caffe.pb.h:24:
In file included from /usr/include/google/protobuf/io/coded_stream.h:130:
In file included from /usr/include/google/protobuf/stubs/common.h:44:
/usr/include/absl/strings/string_view.h:52:26: error: no type named 'string_view' in namespace 'std'
using string_view = std::string_view;
                    ~~~~~^
/usr/include/absl/strings/string_view.h:686:8: error: unknown type name 'string_view'
inline string_view ClippedSubstr(string_view s, size_t pos,
       ^
/usr/include/absl/strings/string_view.h:686:34: error: unknown type name 'string_view'
inline string_view ClippedSubstr(string_view s, size_t pos,
                                 ^
/usr/include/absl/strings/string_view.h:687:45: error: use of undeclared identifier 'string_view'
                                 size_t n = string_view::npos) {
                                            ^
/usr/include/absl/strings/string_view.h:697:11: error: unknown type name 'string_view'
constexpr string_view NullSafeStringView(const char* p) {
          ^
/usr/include/absl/strings/string_view.h:698:14: error: use of undeclared identifier 'string_view'
  return p ? string_view(p) : string_view();
             ^
/usr/include/absl/strings/string_view.h:698:31: error: use of undeclared identifier 'string_view'
  return p ? string_view(p) : string_view();
                              ^
In file included from ncnn/tools/caffe/caffe2ncnn.cpp:19:
In file included from ncnn/build/tools/caffe/caffe.pb.h:24:
In file included from /usr/include/google/protobuf/io/coded_stream.h:130:
In file included from /usr/include/google/protobuf/stubs/common.h:46:
In file included from /usr/include/google/protobuf/stubs/port.h:45:
/usr/include/google/protobuf/port_def.inc:200:1: error: static assertion failed due to requirement '201103L >= 201402L': Protobuf only supports C++14 and newer.
static_assert(PROTOBUF_CPLUSPLUS_MIN(201402L), "Protobuf only supports C++14 and newer.");
^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ncnn/tools/caffe/caffe2ncnn.cpp:19:
In file included from ncnn/build/tools/caffe/caffe.pb.h:24:
In file included from /usr/include/google/protobuf/io/coded_stream.h:130:
In file included from /usr/include/google/protobuf/stubs/common.h:56:
/usr/include/google/protobuf/port_def.inc:200:1: error: static assertion failed due to requirement '201103L >= 201402L': Protobuf only supports C++14 and newer.
static_assert(PROTOBUF_CPLUSPLUS_MIN(201402L), "Protobuf only supports C++14 and newer.");
^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ncnn/tools/caffe/caffe2ncnn.cpp:19:
In file included from ncnn/build/tools/caffe/caffe.pb.h:24:
In file included from /usr/include/google/protobuf/io/coded_stream.h:132:
In file included from /usr/include/absl/log/absl_check.h:38:
In file included from /usr/include/absl/log/internal/check_impl.h:19:
In file included from /usr/include/absl/log/internal/check_op.h:36:
/usr/include/absl/log/internal/nullstream.h:49:32: error: no type named 'string_view' in namespace 'absl'
  NullStream& AtLocation(absl::string_view, int) { return *this; }
                         ~~~~~~^
In file included from ncnn/tools/caffe/caffe2ncnn.cpp:19:
In file included from ncnn/build/tools/caffe/caffe.pb.h:24:
In file included from /usr/include/google/protobuf/io/coded_stream.h:132:
In file included from /usr/include/absl/log/absl_check.h:38:
In file included from /usr/include/absl/log/internal/check_impl.h:19:
In file included from /usr/include/absl/log/internal/check_op.h:37:
In file included from /usr/include/absl/log/internal/strip.h:24:
In file included from /usr/include/absl/log/internal/log_message.h:41:
In file included from /usr/include/absl/log/log_entry.h:35:
In file included from /usr/include/absl/time/time.h:89:
/usr/include/absl/time/civil_time.h:489:27: error: no type named 'string_view' in namespace 'absl'
bool ParseCivilTime(absl::string_view s, CivilSecond* c);
                    ~~~~~~^
/usr/include/absl/time/civil_time.h:490:27: error: no type named 'string_view' in namespace 'absl'
bool ParseCivilTime(absl::string_view s, CivilMinute* c);
                    ~~~~~~^
/usr/include/absl/time/civil_time.h:491:27: error: no type named 'string_view' in namespace 'absl'
bool ParseCivilTime(absl::string_view s, CivilHour* c);
                    ~~~~~~^
/usr/include/absl/time/civil_time.h:492:27: error: no type named 'string_view' in namespace 'absl'
bool ParseCivilTime(absl::string_view s, CivilDay* c);
                    ~~~~~~^
/usr/include/absl/time/civil_time.h:493:27: error: no type named 'string_view' in namespace 'absl'
bool ParseCivilTime(absl::string_view s, CivilMonth* c);
                    ~~~~~~^
/usr/include/absl/time/civil_time.h:494:27: error: no type named 'string_view' in namespace 'absl'
bool ParseCivilTime(absl::string_view s, CivilYear* c);
                    ~~~~~~^
/usr/include/absl/time/civil_time.h:509:34: error: no type named 'string_view' in namespace 'absl'
bool ParseLenientCivilTime(absl::string_view s, CivilSecond* c);
                           ~~~~~~^
```
</details>

https://github.com/Tencent/ncnn/blob/7573faae526cb650faaacf82edc8e48de407afdb/tools/CMakeLists.txt#L2

Setting `CMAKE_CXX_STANDARD` to 17 works.

Please note that the minimum CMake version would have to be bumped to 3.8: https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD.html

**Comment by redthing1** [comment]
Yeah, this is necessary to fix a bunch of issues with absl that I had

**Comment by nihui** [comment]
fixed in https://github.com/Tencent/ncnn/commit/cd1c0d6eab505ce71a4cb1ed6b017609dbb80ffd

**Comment by 414n** [comment]
I doubt that was fixed by cd1c0d6, as I had to apply the workaround described in the first post on version 20250503 to solve the error described.
I have abseil-cpp-20250512.1 and protobuf3-31.1 installed on the system.
